### PR TITLE
Introduce template functions to handle validation errors and non-validation errors

### PR DESCRIPTION
**Pull request by pkong-ds**
ref DEV-1763

## Blocked by
- #4729 

~@louischan-oursky My attempt to implement what we agreed on yesterday, please check if ok!~

### CODE REVIEW

File: pkg/util/template/funcerror.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		return apiErr
+	}
+}
+
+// This function takes any error and returns a validation error
+//
+// i.e. err.Reason == `ValidationFailed`
+//
+// otherwise it returns nil
+func GetValidationError(apiErr *apierrors.APIError, inputFieldName string) (validationErr *apierrors.APIError) {
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror.go:29]: This is not correct. What I told you is to return a validation error with details specific to the given field. So this function should
1. Clone apiErr
2. Filter out any details that is not related to inputFieldName.
3. Return the new apiErr

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/util/template/funcerror.go (hunk 1 above)
Comment (pkong-ds) [reply pkg/util/template/funcerror.go:29]: > Filter out any details that is not related to inputFieldName.

Say given below error and inputFieldName `password_input`, by `filter out any details`, should we remove all other items in `missing`?

## Input apiErr

```
{
	"code": 400,
	"info": {
		"causes": [
			{
				"details": {
					"actual": [
						"gorilla.csrf.Token",
						"x_step"
					],
					"expected": [
						"x_confirm_password",
						"x_password"
					],
					"missing": [
						"x_confirm_password",
						"x_password"
					]
				},
				"kind": "required",
				"location": ""
			}
		]
	},
	"message": "invalid value",
	"name": "Invalid",
	"reason": "ValidationFailed"
}
```

## output

```
{
	"code": 400,
	"info": {
		"causes": [
			{
				"details": {
					"actual": [
						"gorilla.csrf.Token",
						"x_step"
					],
					"expected": [
						"x_confirm_password",
						"x_password"
					],
					"missing": [
						"x_password"
					]
				},
				"kind": "required",
				"location": ""
			}
		]
	},
	"message": "invalid value",
	"name": "Invalid",
	"reason": "ValidationFailed"
}
```

Comment (louischan-oursky) [reply pkg/util/template/funcerror.go:29]: Good question. This is a special case. The keyword `required` reports the error in location one level up. So you're right in this case.

Please also add tests to test this filter function,

Comment (pkong-ds) [reply pkg/util/template/funcerror.go:29]: 2 cases
- Normal case - `location` is non-empty, can just use
- special case - `location` is empty, need to loop through `missing`

```
`/login_id` -> phone
`/x_login_id` -> phone
```

### Proposed api

```
{{ getValidationError apiErr "login_id" "x_login_id" }}
```

**Comment by pkong-ds** [comment]
@louischan-oursky Tried a more declarative api, please help check if okay 🙏

File: pkg/util/template/funcerror_test.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+				"kind": "required",
+				"location": ""
+			}
+		]
+	},
+	"message": "invalid value",
+	"name": "Invalid",
+	"reason": "ValidationFailed"
+}
+`)
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror_test.go:51]: Given

```
{
    "code": 400,
    "info": {
        "causes": [
            {
                "details": {
                    "actual": [
                        "gorilla.csrf.Token",
                        "x_step"
                    ],
                    "expected": [
                        "x_confirm_password",
                        "x_password"
                    ],
                    "missing": [
                        "x_confirm_password",
                        "x_password"
                    ]
                },
                "kind": "required",
                "location": ""
            },
            {
              "kind": "minLength",
              "location": "/x_password"
            },
            {
              "kind": "minLength",
              "location": "/x_confirm_password"
            }
        ]
    },
    "message": "invalid value",
    "name": "Invalid",
    "reason": "ValidationFailed"
}
```

```
resolveError $.RawError (dict
  newPasswordField (dict
    "by_location" (list "x_password")
  )
)
```

it should return

```
{
    "code": 400,
    "info": {
        "causes": [
            {
                "details": {
                    "actual": [
                        "gorilla.csrf.Token",
                        "x_step"
                    ],
                    "expected": [
                        "x_password"
                    ],
                    "missing": [
                        "x_password"
                    ]
                },
                "kind": "required",
                "location": ""
            },
            {
              "kind": "minLength",
              "location": "/x_password"
            }
        ]
    },
    "message": "invalid value",
    "name": "Invalid",
    "reason": "ValidationFailed"
}

```

---

Problems in your proposed API

1. Things should be optional. If I do not care about reason, I do not need to specify `by_reason`
2. You still did not "slice" the errors. See my above expected output.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/util/template/funcerror_test.go (hunk 2 above)
Comment (pkong-ds) [reply pkg/util/template/funcerror_test.go:51]: okok! Lemme update it

**Comment by pkong-ds** [comment]
@louischan-oursky Updated, thanks!

- Added slice (trim) validation errors
- Allow optional fields (even empty dict)

File: pkg/api/apierrors/error.go [hunk 3]
```diff
@@ -65,6 +65,19 @@ func (e *APIError) HasCause(kind string) bool {
 	return false
 }
 
+func (e *APIError) Clone() *APIError {
+	info := make(map[string]interface{})
+	for k, v := range e.Info {
+		info[k] = v
+	}
```
Comment (louischan-oursky) [inline pkg/api/apierrors/error.go:72]: The clone is not deep.

File: pkg/util/template/funcerror.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+func getValidationErrorByLocation(apiErr *apierrors.APIError, locations []string) *apierrors.APIError {
+	out := apiErr.Clone()
+	causes, ok := out.Info["causes"]
+	if !ok {
+		panic("apiErr.Info['causes'] is not set")
+	}
+	causeArr, ok := causes.([]interface{})
+	if !ok {
+		panic("apiErr.Info['causes'] is not an array")
+	}
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror.go:105]: No panic in this function. Just return nil if something is unexpected.

File: pkg/util/template/funcerror.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+func trimValidationErrorCauses(causes []map[string]interface{}, locations []string) []map[string]interface{} {
+	trimmedCauses := slice.Map(causes, func(cause map[string]interface{}) map[string]interface{} {
+		causeLoc, ok := cause["location"]
+		if !ok {
+			panic("cause['location'] is not set")
+		}
+		causeLoc, ok = causeLoc.(string)
+		if !ok {
+			panic("cause['location'] is not string")
+		}
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror.go:135]: Actually, you can do both with `location, ok := case["location"].(string)`

File: pkg/util/template/funcerror.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+	kind, ok := cause["kind"]
+	if !ok {
+		panic("cause['kind'] is not set")
+	}
+	kind, ok = kind.(string)
+	if !ok {
+		panic("cause['kind'] is not string")
+	}
+	if kind != "required" {
+		panic(fmt.Errorf("unexpected cause without location - expected: required, got: %v", kind))
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror.go:172]: Again, do not panic. We are handling error here. Panic in error handling will cause the original error to lose.

Instead we should handle error in a resilient way. I mention we only handle the case that we know, for unknown case, we just do nothing. See what I mean?

- We first check `kind`, instead of `location`.
- If `kind` is `required` and location is `""`, then this is a situation we know how to handle.
- Otherwise, if location is not `""`, then this is another situation we know how to handle.
- Otherwise, we do not panic and return gracefully.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
@louischan-oursky Updated, thanks!

Seems we agree on the API of `resolveError` already. Lemme start applying the changes across entire authUI 

Question - is authUI v1 needed? I plan to apply on v2 only 🙏

**Comment by pkong-ds** [comment]
Used `resolveError` and new `__render_error.html` in below pages

<details>

<summary> Expand me </summary>


- `new_password_field.html`
  - `reset_password.html`
  - `change_password.html`
  - `create_password.html`
- `__password_field.html`
  - `enter_password.html`
  - `ldap_login.html`

</details>

I guess this PR remaining actionables

1. Ensure all changed screens errors work normally; that is, to attach screenshot videos here (or add tests in https://github.com/oursky/e2e-test-authgear? not sure if the effort is of similar amount)
2. Change other screens that requires refactoring
    - quick search seems remaining OAuthError that has hacky error usage

Actionables not expected in this PR

1. Replace all `__error.html` with `__render_error.html`; that is, pass error explicitly to render

Let's sync tmr 🙏

File: resources/authgear/templates/en/web/authflowv2/__render_error.html [hunk 7]
```diff
@@ -0,0 +1,306 @@
+{{/*
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__render_error.html:1]: Remove this. We do not want to maintain __error.html and __render_error.html

File: resources/authgear/templates/en/web/authflowv2/change_password.html [hunk 8]
```diff
@@ -1,9 +1,29 @@
...
+
+{{ $has_unknown_err := not (isNil $unknown_err )}}
 {{ $unknown_error_message := "" }}
-{{ if $.HasUnknownError }}
-  {{ $unknown_error_message = (include "authflowv2/__error.html" .) }}
+{{ if $has_unknown_err }}
+  {{ $unknown_error_message = (include "authflowv2/__render_error.html" (dict
+    "Ctx" $
+    "Error" $unknown_err
+  )) }}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/change_password.html:26]: You actually do not need __render_error.html, you can just write

```
(include "authflowv2/__error.html" (merge (dict "Error" thenewerror) .))
```

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
Updated, thanks

**Comment by pkong-ds** [comment]
As per offline discussion, This PR 

### Will
✅ make commit history cleaner 
⏳ look for other problematic pattern (see [Comment on problematic patten](https://github.com/authgear/authgear-server/pull/4705/commits/e1b71887efab2e66cb925cf1d055db9c0466b391), wip)

### Won't
⛔ refactor `$oauth_error` handling in `login.html` and `signup.html`

File: resources/authgear/templates/en/web/authflowv2/ldap_login.html [hunk 9]
```diff
@@ -64,8 +92,7 @@ <h1 class="screen-title">
       "Type" "old-password"
       "AutoFocus" $.ShouldFocusInput
       "Classname" "w-full"
-      "HasError" $.PasswordInputError.HasError
-      "HasErrorMessage" $.PasswordInputError.HasErrorMessage
+      "Error" $pw_err
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: I take a closer at this, and I found after your change, the behavior is different. Have you ever tried the original behavior?

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: resources/authgear/templates/en/web/authflowv2/ldap_login.html (hunk 9 above)
Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: Right, missed that `InvalidCredentials` should be 

1. `both username and password input` red border
2. no error messages for input fields
3.  show alert message by `Invalid credentials`

Lemme fix that 🙏

Comment (louischan-oursky) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: So as I said, you do not need to change the interface of __password_field.html

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: >  do not need to change the interface of __password_field.html
Disagree on this. I think we need `.Error` to be passed into `__password_field.html` for below

```
<p class="input__error-message" data-text-field-target="errorMessage">
      {{ template "authflowv2/__error.html" (merge (dict "Error" $.Error) $.Ctx) }}
    </p>
```

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: So I suggest below diff

```diff
- "HasError`
+ "Error"
   "HasErrorMessage" // keep

```

Comment (louischan-oursky) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: OK. Remove .HasErrorMessage and add .ErrorMessage

Then in __password_field.html, render .ErrorMessage directly.

Comment (louischan-oursky) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: I want it to support two things

- A flag (or a prop) to make it in red
- A prop to pass it an error message to render

So this becomes

- HasError
- ErrorMessage

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: Sounds great. Lemme implement that.

Will do similarly for `__new_password_field.html` 🙏 Thanks daic

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: Sor @louischan-oursky , on second thought, why not pass the entire Error over HasError flag?

So it becomes

- Error
- ErrorMessage

Comment (louischan-oursky) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: > Have you ever tried the original behavior?

Did you?

If you have, then you will see in a case, the field is just in red, without any error message.

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: > Did you?

Yes I did. I understand that we want 2 separate controls on `field red` and `has error message`

I am suggesting `Error` over `HasError` because that way the consumer `__password_field.html` has control over the entire Error object, instead of a processed flag.

It's rather trivial, but just wanna point it out

Comment (louischan-oursky) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: The rationale for __password_field.html to consume an entire Error object is __password_field.html needs to do something with .Error. But it does not need to. It just need to display an given error message.

Or you found such a case?

Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/ldap_login.html:95]: > Or you found such a case?

Nope. Maybe let's just pass it when it's needed (which is unlikely).

Will implement below interface 

- HasError
- ErrorMessage

for new_password_field, will be

```
"HasNewPasswordError" $has_pw_err
    "NewPasswordErrorMessage" $pw_error_message
    "HasConfirmPasswordError" $confirm_pw_err
    "ConfirmPasswordErrorMessage" $confirm_pw_error_message
```

**Comment by pkong-ds** [comment]
Updated, thanks! Changes since [Comment on problematic patten](https://github.com/authgear/authgear-server/pull/4705/commits/e7a8541afcfc79603f19c151e699f78a400b68b5)

Will squash commits after finalizing

**Comment by pkong-ds** [comment]
## change_password.html 



## create_password.html 



## enter_oob_otp.html 



## enter_password.html 



## enter_recovery_code.html 



## enter_totp.html 



## forgot_password_otp.html 



## Idap_login.html 

https://github.com/user-attachments/assets/8c08ae3b-ba89-4190-9f1b-d430a3107aeb


## reset_password.html 



## setup_oob_otp.html 



## setup_totp_verify.html

File: pkg/util/template/funcerror.go [hunk 10]
```diff
@@ -93,17 +93,16 @@ func getError(apiErr *apierrors.APIError, getErrInput *GetErrorInput) *apierrors
 	if apiErr == nil {
 		return nil
 	}
+	apiErrByReason := getErrorByReason(apiErr, getErrInput.ByReason)
+	if apiErrByReason != nil {
+		return apiErrByReason
+	}
```
Comment (louischan-oursky) [inline pkg/util/template/funcerror.go:99]: I thought by_location and by_reason are independent. But in your implementation they are not.

File: resources/authgear/templates/en/web/authflowv2/__password_field.html [hunk 11]
```diff
@@ -23,9 +33,9 @@
     "PasswordRules" .PasswordRules
     "InputAttrs" $input_attrs
   ) }}
-  {{ if $has_err }}
+  {{ if $.HasError }}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__password_field.html:36]: Checks ErrorMessage instead of HasError. So technically, it can display an error message without being highlighted in red.

File: resources/authgear/templates/en/web/authflowv2/__new_password_field.html [hunk 12]
```diff
@@ -36,13 +36,12 @@
         `
       )
     }}
-    {{ $has_pw_input_err_msg := $has_new_pw_err }}
     {{ $pw_input_error_class := "input__error-message"}}
-    {{ if not $has_pw_input_err_msg }}
+    {{ if (eq $.NewPasswordErrorMessage "") }}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__new_password_field.html:40]: `{{ if $.NewPasswordErrorMessage }}`

File: resources/authgear/templates/en/web/authflowv2/__new_password_field.html [hunk 13]
```diff
@@ -61,20 +60,15 @@
...
       )
     }}
-    {{/* confirm password has error msg only if 
-      - confirm password input has error; AND
-      - new password input has no error
-     */}}
-    {{ $has_confirm_password_input_err_msg := (and $has_confirm_pw_err (not $has_new_pw_err))}}
     {{ $confirm_input_error_class := "input__error-message"}}
-    {{ if (not $has_confirm_password_input_err_msg) }}
+    {{ if (eq $.ConfirmPasswordErrorMessage "") }}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/__new_password_field.html:64]: `{{ if $.ConfirmPasswordErrorMessage }}`

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/util/template/funcerror.go (hunk 10 above)
Comment (pkong-ds) [reply pkg/util/template/funcerror.go:99]: > I thought by_location and by_reason are independent. 

Should they be independent? 

What if the caller wants to match `ValidationFailed` without providing any location, as seen in `setup_oob_otp.html`? 🙏

File: resources/authgear/templates/en/web/authflowv2/__new_password_field.html (hunk 12 above)
Comment (pkong-ds) [reply resources/authgear/templates/en/web/authflowv2/__new_password_field.html:40]: Will update as `{{ if (not $.NewPasswordErrorMessage) }}`
Quoting [golang doc](https://pkg.go.dev/text/template#hdr-Actions) here for reference 🙏 

```
{{if pipeline}} T1 {{end}}
	If the value of the pipeline is empty, no output is generated;
	otherwise, T1 is executed. The empty values are false, 0, any
	nil pointer or interface value, and any array, slice, map, or
	string of length zero.
	Dot is unaffected.
```

**Comment by pkong-ds** [comment]
Updated, thanks!

**Review by louischan-oursky — APPROVED** [review]

File: pkg/util/template/funcerror.go (hunk 10 above)
Comment (pkong-ds) [reply pkg/util/template/funcerror.go:99]: ok, thanks daic!
